drivers/dcf77/dcf77.c

```diff
@@ -0,0 +1,257 @@
+                    dev->bitseq[dev->bitCounter] = 1;
+                }
+                else {
+                    dev->bitseq[dev->bitCounter] = 0;
+
+                }
+
+                dev->bitCounter++;
+                if (dev->bitCounter >= DCF77_READING_CYCLE) {
+                    mutex_unlock(&dev->event_lock);
+                    dev->bitCounter = 0;
+                    dev->startTime = xtimer_now_usec();
+                    dev->internal_state = DCF77_STATE_START;
+                }
+
+            }
+            break;
+
+    }
+
+}
+
+
+
+
+/**
+ * @brief   Initalize the Device
+ *
+ * @param   dev     device
+ * @param   params  device_params
+ *
+ * @retval  DCF77_OK             Success
+ * @retval  DCF77_INIT_ERROR     Error in initalization
+ */
+
+int dcf77_init(dcf77_t *dev, const dcf77_params_t *params)
+{
+    DEBUG("dcf77_init\n");
+
+    /* check parameters and configuration */
+    assert(dev && params);
+    dev->params = *params;
+    dev->internal_state = DCF77_STATE_IDLE;
+    dev->bitCounter = 0;
+    gpio_init_int(dev->params.pin, dev->params.in_mode, GPIO_BOTH, _level_cb,
+                  dev);
+
+    mutex_init(&dev->event_lock);
+    return DCF77_OK;
+}
+
+/**
+ * @brief   Formats the information after all bits of a cycle have been received
+ *
+ * @param   dev               device
+ * @param   time              datastruct for timeinformation
+ *
+ * @retval  DCF77_OK          Success
+ * @retval  DCF77_TIMEOUT     Timeout occurred before level was reached
+ * @retval  DCF77_NOCSUM      Parity Bits arent correct
+ */
+
+int dcf77_read(dcf77_t *dev, struct tm *time)
+{
+    assert(dev);
+    mutex_lock(&dev->event_lock);
+
+    if ((dev->bitseq[DCF77_MESZ1] == 1) && (dev->bitseq[DCF77_MESZ2] == 0)) {
+        time->tm_isdst = 1;
+    }
+    else {
+        time->tm_isdst = 0;
+    }
+
+    uint8_t minute = bcd2dez(dev->bitseq[DCF77_MINUTE_EIGHT],
+                             dev->bitseq[DCF77_MINUTE_QUAD],
+                             dev->bitseq[DCF77_MINUTE_SECOND],
+                             dev->bitseq[DCF77_MINUTE_SINGLE]);
+    minute = minute + POTENZY10 * bcd2dez(0, dev->bitseq[DCF77_MINUTE_FOURTIES],
+                                          dev->bitseq[DCF77_MINUTE_TEWENTIES],
+                                          dev->bitseq[DCF77_MINUTE_TENNER]);
+
+    if (checkEvenParity(DCF77_MINUTE_SINGLE, DCF77_MINUTE_FOURTIES,
+                        dev) != dev->bitseq[DCF77_MINUTE_PR]) {
+        return DCF77_NOCSUM;
+    }
+
+    time->tm_min = minute;
+
+    uint8_t hour = bcd2dez(dev->bitseq[DCF77_HOUR_EIGHT], dev->bitseq[DCF77_HOUR_QUAD],
+                           dev->bitseq[DCF77_HOUR_SECOND], dev->bitseq[DCF77_HOUR_SINGLE]);
+    hour = hour + POTENZY10 * bcd2dez(0, 0, dev->bitseq[DCF77_HOUR_TEWENTIES],
+                                      dev->bitseq[DCF77_HOUR_TENNER]);
+
+    if (checkEvenParity(DCF77_HOUR_SINGLE, DCF77_HOUR_TEWENTIES,
```

Comment:
[`__builtin_parityll()`](https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html) exists and `__builtin_parity()` will also just be replaced by a [library function](https://github.com/intel/linux-sgx/blob/master/sdk/compiler-rt/paritysi2.c) on most platforms, but let's play it save and keep things 32 bit if we don't need the full width.  

`__builtin_parity((dev->bitseq.bits >> DCF77_HOUR_SHIFT) & (DCF77_HOUR_MASK >> DCF77_HOUR_SHIFT))`